File: routes/graph.js
const express = require('express');
const path = require('path');
const fs = require('fs');
const { execFile } = require('child_process');
const util = require('util');
const execFileAsync = util.promisify(execFile);

const { discoverNamespaces, getItemsFromSnapshot, buildGraph } = require('../utils/graph-builder');

const router = express.Router();

// --- Realtime (kubectl) helpers ---

async function execKubectl(args) {
  try {
    const argList = args.split(/\s+/);
    const { stdout } = await execFileAsync('kubectl', argList, {
      encoding: 'utf8',
      timeout: 30000,
      maxBuffer: 50 * 1024 * 1024,
    });
    const bytes = Buffer.byteLength(stdout, 'utf8');
    const parsed = JSON.parse(stdout);
    console.log(`[graph] kubectl ${args.split(' -')[0]}: ${(bytes / 1024).toFixed(1)}KB, ${parsed.items?.length ?? 0} items`);
    return parsed;
  } catch (e) {
    console.warn(`[graph] kubectl ${args}: ${e.message?.split('\n')[0]}`);
    return { items: [] };
  }
}

async function fetchLiveData() {
  const batches = [
    { resources: 'deployments,statefulsets,daemonsets,cronjobs', keys: ['deployments', 'statefulsets', 'daemonsets', 'cronjobs'] },
    { resources: 'services,configmaps,ingresses', keys: ['services', 'configmaps', 'ingresses'] },
    { resources: 'secrets,serviceaccounts,rolebindings', keys: ['secrets', 'serviceaccounts', 'rolebindings'] },
    { resources: 'pods', keys: ['pods'] },
    { resources: 'hpa', keys: ['horizontalpodautoscalers'] },
    { resources: 'pvc', keys: ['persistentvolumeclaims'] },
  ];

  const optionalBatches = [
    { resources: 'gateways.gateway.networking.k8s.io', keys: ['gateways'] },
    { resources: 'httproutes.gateway.networking.k8s.io', keys: ['httproutes'] },
    { resources: 'tcproutes.gateway.networking.k8s.io', keys: ['tcproutes'] },
  ];

  const nsData = new Map();
  const allNamespaces = new Set();

  function ingest(data, keys) {
    const items = data?.items || [];
    const kindToKey = {};
    for (const key of keys) {
      const kindMap = {
        'deployments': 'Deployment',
        'statefulsets': 'StatefulSet',
        'daemonsets': 'DaemonSet',
        'cronjobs': 'CronJob',
        'services': 'Service',
        'configmaps': 'ConfigMap',
        'ingresses': 'Ingress',
        'secrets': 'Secret',
        'serviceaccounts': 'ServiceAccount',
        'rolebindings': 'RoleBinding',
        'pods': 'Pod',
        'horizontalpodautoscalers': 'HorizontalPodAutoscaler',
        'persistentvolumeclaims': 'PersistentVolumeClaim',
        'gateways': 'Gateway',
        'httproutes': 'HTTPRoute',
        'tcproutes': 'TCPRoute',
      };
      if (kindMap[key]) kindToKey[kindMap[key]] = key;
    }

    for (const item of items) {
      const ns = item.metadata?.namespace || '_cluster';
      const kind = item.kind;
      const key = kindToKey[kind] || keys[0];

      allNamespaces.add(ns);
      if (!nsData.has(ns)) nsData.set(ns, new Map());
      const nsMap = nsData.get(ns);
      if (!nsMap.has(key)) nsMap.set(key, []);
      nsMap.get(key).push(item);
    }
  }

  const allBatches = [...batches, ...optionalBatches];
  const results = await Promise.all(
    allBatches.map(batch => execKubectl(`get ${batch.resources} -A -o json`))
  );

  for (let i = 0; i < allBatches.length; i++) {
    ingest(results[i], allBatches[i].keys);
  }

  return { nsData, namespaces: [...allNamespaces] };
}

// GET /api/graph
router.get('/graph', async (req, res) => {
  const isSnapshot = req.query.snapshot === 'true';

  try {
    if (isSnapshot) {
      const rootDir = path.join(__dirname, '..');
      const localBackup = path.join(rootDir, 'k8s-snapshot');
      const fallbackPath = process.env.K8S_SNAPSHOT_PATH || localBackup;

      let dataPaths;
      if (req.query.path) {
        dataPaths = [path.resolve(req.query.path)];
      } else if (fs.existsSync(localBackup)) {
        dataPaths = [localBackup];
      } else {
        dataPaths = [fallbackPath];
      }

      const namespaceDirs = discoverNamespaces(dataPaths);
      const namespaceList = [...namespaceDirs.keys()];

      const getItemsFn = (ns, resourceKey) => {
        const nsDir = namespaceDirs.get(ns);
        if (!nsDir) return [];
        return getItemsFromSnapshot(nsDir, resourceKey);
      };

      res.json(buildGraph(getItemsFn, namespaceList));
    } else {
      const { nsData, namespaces } = await fetchLiveData();

      const getItemsFn = (ns, resourceKey) => {
        const nsMap = nsData.get(ns);
        if (!nsMap) return [];
        return nsMap.get(resourceKey) || [];
      };

      res.json(buildGraph(getItemsFn, namespaces));
    }
  } catch (err) {
    console.error('[graph] Error:', err.message);
    res.status(500).json({ error: 'Failed to fetch graph data' });
  }
});

module.exports = router;

File: cmd/server/routes/graph.go
package routes

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os/exec"
	"sync"
	"time"

	"kubecmds-viz/server/graph"
	"kubecmds-viz/server/store"
)

func registerGraph(mux *http.ServeMux) {
	mux.HandleFunc("GET /api/graph", handleGraph)
}

// GET /api/graph
// Snapshot mode: reads from k8s-snapshot/ YAML files.
// Realtime mode: runs 9 parallel kubectl get -A -o json batches.
func handleGraph(w http.ResponseWriter, r *http.Request) {
	if r.URL.Query().Get("snapshot") == "true" {
		handleGraphSnapshot(w)
		return
	}
	handleGraphRealtime(w, r)
}

// --- Snapshot ---

func handleGraphSnapshot(w http.ResponseWriter) {
	namespaces := store.ListBackupNamespaces()
	getItems := func(ns, resourceKey string) []graph.K8sItem {
		list := store.LoadYaml(resourceKey+".yaml", ns)
		if list == nil {
			return nil
		}
		return list.Items
	}
	result := graph.BuildGraph(getItems, namespaces)
	writeJSON(w, http.StatusOK, result)
}

// --- Realtime ---

type batchSpec struct {
	resources string
	keys      []string
}

var coreBatches = []batchSpec{
	{"deployments,statefulsets,daemonsets,cronjobs", []string{"deployments", "statefulsets", "daemonsets", "cronjobs"}},
	{"services,configmaps,ingresses", []string{"services", "configmaps", "ingresses"}},
	{"secrets,serviceaccounts,rolebindings", []string{"secrets", "serviceaccounts", "rolebindings"}},
	{"pods", []string{"pods"}},
	{"hpa", []string{"horizontalpodautoscalers"}},
	{"pvc", []string{"persistentvolumeclaims"}},
}

var optionalBatches = []batchSpec{
	{"gateways.gateway.networking.k8s.io", []string{"gateways"}},
	{"httproutes.gateway.networking.k8s.io", []string{"httproutes"}},
	{"tcproutes.gateway.networking.k8s.io", []string{"tcproutes"}},
}

// kindToKey maps K8s Kind → resourceKey used in getItemsFn.
var kindToKey = map[string]string{
	"Deployment":              "deployments",
	"StatefulSet":             "statefulsets",
	"DaemonSet":               "daemonsets",
	"CronJob":                 "cronjobs",
	"Service":                 "services",
	"ConfigMap":               "configmaps",
	"Ingress":                 "ingresses",
	"Secret":                  "secrets",
	"ServiceAccount":          "serviceaccounts",
	"RoleBinding":             "rolebindings",
	"Pod":                     "pods",
	"HorizontalPodAutoscaler": "horizontalpodautoscalers",
	"PersistentVolumeClaim":   "persistentvolumeclaims",
	"Gateway":                 "gateways",
	"HTTPRoute":               "httproutes",
	"TCPRoute":                "tcproutes",
}

type batchResult struct {
	items []graph.K8sItem
	err   string
}

func handleGraphRealtime(w http.ResponseWriter, r *http.Request) {
	allBatches := append(coreBatches, optionalBatches...)
	results := make([]batchResult, len(allBatches))

	var wg sync.WaitGroup
	for i, batch := range allBatches {
		wg.Add(1)
		go func(i int, b batchSpec) {
			defer wg.Done()
			results[i] = fetchBatch(b, r)
		}(i, batch)
	}
	wg.Wait()

	// All core batches failed → kubectl is broken.
	coreFailures := 0
	for i := range coreBatches {
		if results[i].err != "" {
			coreFailures++
		}
	}
	if coreFailures == len(coreBatches) {
		writeJSON(w, http.StatusInternalServerError, map[string]any{
			"message": results[0].err,
		})
		return
	}

	// Build ns → resourceKey → []K8sItem index.
	nsData := map[string]map[string][]graph.K8sItem{}
	allNsSet := map[string]bool{}

	for i, batch := range allBatches {
		for _, item := range results[i].items {
			meta, _ := item["metadata"].(map[string]interface{})
			ns, _ := meta["namespace"].(string)
			if ns == "" {
				ns = "_cluster"
			}
			kind, _ := item["kind"].(string)
			key, ok := kindToKey[kind]
			if !ok {
				key = batch.keys[0]
			}

			allNsSet[ns] = true
			if nsData[ns] == nil {
				nsData[ns] = map[string][]graph.K8sItem{}
			}
			nsData[ns][key] = append(nsData[ns][key], item)
		}
	}

	namespaces := make([]string, 0, len(allNsSet))
	for ns := range allNsSet {
		namespaces = append(namespaces, ns)
	}

	getItems := func(ns, resourceKey string) []graph.K8sItem {
		if m, ok := nsData[ns]; ok {
			return m[resourceKey]
		}
		return nil
	}

	result := graph.BuildGraph(getItems, namespaces)
	writeJSON(w, http.StatusOK, result)
}

// fetchBatch runs a single kubectl get batch and returns parsed items.
func fetchBatch(b batchSpec, r *http.Request) batchResult {
	args := []string{"get", b.resources, "-A", "-o", "json"}
	cmd := exec.Command("kubectl", args...)
	cmd.WaitDelay = 30 * time.Second

	out, err := cmd.Output()
	if err != nil {
		msg := err.Error()
		if ee, ok := err.(*exec.ExitError); ok {
			msg = string(ee.Stderr)
		}
		log.Printf("[graph] kubectl get %s: %s", b.resources, msg)
		return batchResult{err: msg}
	}

	var list struct {
		Items []graph.K8sItem `json:"items"`
	}
	if err := json.Unmarshal(out, &list); err != nil {
		return batchResult{err: fmt.Sprintf("parse error: %v", err)}
	}

	log.Printf("[graph] kubectl get %s: %d items (%.1fKB)",
		b.resources, len(list.Items), float64(len(out))/1024)

	return batchResult{items: list.Items}
}
